### PR TITLE
Major refactor of ImagesGallery to use cleaner models and events

### PR DESCRIPTION
**Pull request by brianhelba**

### CODE REVIEW

**Comment by brianhelba** [comment]
@ronichoudhury @msmolens I'll talk to you both in person about these changes. Feel free to check out and test the branch, but I would recommend not looking at the code deltas until we have a chance to walk through it together.

There's even more code that can be removed here, but it will require webpack first.

File: web_external/ImagesGallery/ImagesFilters.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        this._filters = completeFacets.chain()
+            .map(function (completeFacet) {
+                var facetId = completeFacet.id;
+                var facetBins = completeFacet.get('bins');
+
+                var facetFilters = _.chain(facetBins)
+                    .map(function (facetBin) {
+                        var binLabel = facetBin.label;
+                        // Default to all bins included
+                        var binIncluded = true;
```
Comment (ronichoudhury) [inline web_external/ImagesGallery/ImagesFilters.js]: Changing this to `false` should "just work" but it doesn't.

File: web_external/ImagesGallery/ImagesFilters.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        var excludedBinLabels = _.chain(facetFilters)
+            .pick(function (binIncluded, binLabel) {
+                return binIncluded === false;
+            })
+            .keys()
+            .map(isic.SerializeFilterHelpers._stringToHex)
+            .value();
+        // TODO: Could use "facetId + ' in ' + includedBinLabels"" if most are excluded
+        return '(' +
+            isic.SerializeFilterHelpers._stringToHex(facetId) +
```
Comment (ronichoudhury) [inline web_external/ImagesGallery/ImagesFilters.js]: Putting quote characters around the identifier should allow us to simplify this code (won't need `_stringToHex()` anymore hopefully).

File: web_external/ImagesGallery/ImagesFilters.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        return '(' + filterExpression + ')';
+    },
+
+    _categoricalFacetFilterExpression: function (facetId, facetFilters) {
+        var excludedBinLabels = _.chain(facetFilters)
+            .pick(function (binIncluded, binLabel) {
+                return binIncluded === false;
+            })
+            .keys()
+            .map(isic.SerializeFilterHelpers._stringToHex)
```
Comment (ronichoudhury) [inline web_external/ImagesGallery/ImagesFilters.js]: This does always have to be "hexified" because the labels come from arbitrary data sources (i.e. we *don't* control the content).

File: web_external/ImagesGallery/ImagesFilters.js (hunk 2 above)
Comment (ronichoudhury) [inline web_external/ImagesGallery/ImagesFilters.js]: Although... it looks like everything is *de*hexified at the end no matter what, so that may cause problems if this part isn't hexified.

**Comment by ronichoudhury** [comment]
There's a small bug in query.pegjs - the `not` expression should allow whitespace ("`_`") after the opening parenthesis.

File: web_external/ImagesGallery/ImagesView.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+            this.images.fetch();
+
+            this.filters.initialize(collection);
+        });
+        this.listenTo(this.filters, 'change', function () {
+            // TODO: Ideally, this should select and cancel only the old set of requests; until
+            // that's fixed, hopefully nobody will click histogram buttons too early in the page
+            // load
+            girder.cancelRestRequests();
```
Comment (ronichoudhury) [inline web_external/ImagesGallery/ImagesView.js]: Replace this with a mechanism that saves the two requests made (a few lines below) and only cancel those. This mechanism can break the application because there may be other pending requests that should *not* be canceled by action here.

**Comment by ronichoudhury** [comment]
Push `$rolloverImageFilter` and brethren down into styl file where they are used.

**Comment by ronichoudhury** [comment]
Delete rangeSet.js

File: web_external/ImagesGallery/ImagesFilters.js [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        if (fullExpression) {
+            var ast = isic.SerializeFilterHelpers.astParser.parse(fullExpression);
+            ast = isic.SerializeFilterHelpers._dehexify(ast);
+            return isic.SerializeFilterHelpers._specifyAttrTypes(ast);
+        } else {
+            return undefined;
+        }
+    },
+
+    _rangeFacetFilterExpression: function (facetId, facetFilters) {
```
Comment (msmolens) [inline web_external/ImagesGallery/ImagesFilters.js]: - I see errors in this function when toggling the "unknown" groups for "Approximate Age" and "Clinical Size - Longest Diameter (mm)":
    ```
    ImagesFilters.js:100 Uncaught TypeError: Cannot read property '1' of null
    ```

- Toggling "unknown" for "Benign or Malignant" doesn't update the images in the gallery.

- "unknown" handled correctly under "Lesion Diagnosis" either. If I uncheck all the checkboxes, there are still 312 images shown--the exact number of images in the "unknown" category.

- The live site (sometimes) shows more information under "Lesion Diagnosis" (e..g "nevus (3014/12265)").  It looks like this branch always displays only the total.

Comment (brianhelba) [inline web_external/ImagesGallery/ImagesFilters.js]: The first 3 issues here were caused by an incorrect handling of null values. See https://github.com/ImageMarkup/isic-archive/pull/365/commits/c9702d496ea787be3356a4295184eeca56055dd8 for a fix.

**Comment by brianhelba** [comment]
@ronichoudhury PTAL @ https://github.com/ImageMarkup/isic-archive/pull/365/commits/2721949b131e7f8ee16afbe86f221b511df215b1

File: web_external/ImagesGallery/ImagesFilters.js (hunk 5 above)
Comment (brianhelba) [inline web_external/ImagesGallery/ImagesFilters.js]: Edited my last comment.

File: web_external/ImagesGallery/ImagesFilters.js [hunk 6]
```diff
@@ -101,6 +106,7 @@ _.extend(isic.collections.ImagesFilters.prototype, Backbone.Events, {
                 var highBound = parseFloat(rangeMatches[2]);
                 return [lowBound, highBound];
             })
+            .sortBy(_.identity)
```
Comment (ronichoudhury) [inline web_external/ImagesGallery/ImagesFilters.js]: What effect does this have?

File: web_external/ImagesGallery/ImagesFilters.js [hunk 7]
```diff
@@ -131,10 +137,14 @@ _.extend(isic.collections.ImagesFilters.prototype, Backbone.Events, {
                     ')';
                 return '(' + lowBoundExpression + ' or ' + highBoundExpression + ')';
             })
-            .value()
-            // Combine all expressions
-            .join(' and ');
-        return '(' + filterExpression + ')';
+            .value();
+        if (facetFilters['null'] === false) {
```
Comment (ronichoudhury) [inline web_external/ImagesGallery/ImagesFilters.js]: What's the idea in this conditional block? To filter out all null values if the caller asks for that?

File: web_external/ImagesGallery/ImagesFilters.js (hunk 5 above)
Comment (msmolens) [inline web_external/ImagesGallery/ImagesFilters.js]: Nice, the "unknown"/nulls work now.

**Comment by brianhelba** [comment]
@ronichoudhury As discussed in our live conversation, note that the changes to `null` handling in serialization of range facet filters have been overwritten by subsequent changes to not use `null` values at all (instead using `'__null__'`. Sorry for the confusion.

**Comment by ronichoudhury** [comment]
No worries - makes sense.

**Comment by brianhelba** [comment]
@msmolens Ok, the last issue should be fixed. This was more complicated, as I realized that the "ZIP" of completeFacetBins and filteredFacetBins wasn't staying updated as filteredFacet was re-fetched (since the nested bins arrays were getting overwritten with new references).

So, I:
* rewrote the `ImagesFacetView` classes to ZIP up completeFacetBins and filteredFacetBins on demand before every rerender
* rewrote the `_zipFacetBins` function to run in linear time
* removed the use of D3 from `ImagesFacetCategoricalView` (so now only the code related to the `ImagesFacetHistogramView` requires D3)
* removed the need for `ImagesFacetCategoricalView._rerenderSelections` (which should be fast, since it needs to be responsive to user clicks) to use the ZIPed bins

This last change then broke the way that `null` bins were being used, as a `null` value cannot be stored as a data attribute on a DOM element. So, to deal with that problem, and the fact that the use of `null` as an object key was also awkward, I rewrote the `ImagesFacetCollection` to always convert bin labels of `null` to `'__null__'` as soon as they're parsed from the REST request, and use `'__null__'` internally everywhere.

In short, (I believe) it should now fully work as intended. Please test and let me know.

**Comment by msmolens** [comment]
@brianhelba I tested again and don't see any problems. Great work!

**Comment by brianhelba** [comment]
@msmolens Cool, thanks. Want to give this a review so that I can merge?
